File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/model/ContribFeatureTest.java
package org.totschnig.myexpenses.test.model;

import junit.framework.Assert;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.di.AppComponent;
import org.totschnig.myexpenses.model.ContribFeature;
import org.totschnig.myexpenses.preference.PrefHandler;
import org.totschnig.myexpenses.util.licence.LicenceHandler;

import static org.totschnig.myexpenses.model.ContribFeature.USAGES_LIMIT;

public class ContribFeatureTest extends ModelTest {

  public void testRecordUsage() {
    ContribFeature feature = ContribFeature.ATTACH_PICTURE;
    final AppComponent appComponent = ((MyApplication) getContext().getApplicationContext()).getAppComponent();
    LicenceHandler licenceHandler = appComponent.licenceHandler();
    PrefHandler prefHandler = appComponent.prefHandler();
    Assert.assertEquals(USAGES_LIMIT, feature.usagesLeft(prefHandler));
    licenceHandler.setLockState(true);
    feature.recordUsage(prefHandler, licenceHandler);
    Assert.assertEquals(USAGES_LIMIT - 1, feature.usagesLeft(prefHandler));
    licenceHandler.setLockState(false);
    feature.recordUsage(prefHandler, licenceHandler);
    Assert.assertEquals(USAGES_LIMIT - 1, feature.usagesLeft(prefHandler));
    licenceHandler.setLockState(true);
    feature.recordUsage(prefHandler, licenceHandler);
    Assert.assertEquals(USAGES_LIMIT - 2, feature.usagesLeft(prefHandler));
  }
}

File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/model/ContribFeatureTest.kt
package org.totschnig.myexpenses.test.model

import androidx.test.platform.app.InstrumentationRegistry
import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.totschnig.myexpenses.TestApp
import org.totschnig.myexpenses.model.ContribFeature
import org.totschnig.myexpenses.testutils.MockLicenceHandler
import org.totschnig.myexpenses.util.licence.LicenceHandler
import java.time.Duration

class ContribFeatureTest {

    @Test
    fun testRecordUsage() {
        val feature = ContribFeature.DISTRIBUTION
        val application = InstrumentationRegistry.getInstrumentation().targetContext.applicationContext as TestApp
        val licenceHandler = application.licenceHandler as MockLicenceHandler
        assertThat(licenceHandler.usagesLeft(feature)).isTrue()
        application.advanceClock(Duration.ofDays(LicenceHandler.TRIAL_DURATION_DAYS))
        assertThat(licenceHandler.usagesLeft(feature)).isTrue()
        licenceHandler.recordUsage(feature)
        assertThat(licenceHandler.usagesLeft(feature)).isTrue()
        application.advanceClock(Duration.ofDays(LicenceHandler.TRIAL_DURATION_DAYS))
        assertThat(licenceHandler.usagesLeft(feature)).isFalse()
    }
}